### PR TITLE
Implement `@PreMatching` request filter methods

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Comment by dstepanov** [comment]
@yawkat How to fix the type polution check?

**Comment by yawkat** [comment]
Wouldn't it be easier to just run the FilterRunner twice, once before once after routing?

**Comment by yawkat** [comment]
The type pollution is caused by this RouteInfo stuff being moved before the conditional filter routes check: https://github.com/micronaut-projects/micronaut-core/blob/634b3b256da680114cd0a39174656ae70980dfc7/router/src/main/java/io/micronaut/web/router/DefaultRouter.java#L588

**Comment by dstepanov** [comment]
> Wouldn't it be easier to just run the FilterRunner twice, once before once after routing?

Probably not, this change also eliminates multiple invocation of filtering for status routes etc.

**Comment by yawkat** [comment]
I don't really like how specific the FilterRunner is to the server now

**Comment by yawkat** [comment]
Is it maybe possible to make the routing decision a request filter and insert it between the pre- and post-match filters? I'm not a fan of the api changes to FilterRunner

**Comment by dstepanov** [comment]
@yawkat Please check the latest version

**Comment by graemerocher** [comment]
Can someone enlighten me what the problem is we are trying to solve here?

**Comment by dstepanov** [comment]
Just to align with JaxRs. You can also have a filter that will change the request to hit a different route.

**Comment by graemerocher** [comment]
ok are we adding an API to achieve that without JAX-RS? If so can we document it?

**Comment by dstepanov** [comment]
It is documented. The is a new PreMatching annotation.

**Comment by graemerocher** [comment]
ok but how does one change the route?

**Comment by yawkat** [comment]
I think you'd have to change the request path.

File: http/src/main/java/io/micronaut/http/filter/FilterRunner.java [hunk 1]
```diff
@@ -163,6 +252,16 @@ public final ExecutionFlow<HttpResponse<?>> run(HttpRequest<?> request,
         return flow.flatMap(context -> filterResponse(context, iterator, null));
     }
 
+    private List<InternalHttpFilter> filterFilters(List<InternalHttpFilter> filters, HttpRequest<?> request) {
+        List<InternalHttpFilter> filtersToRun = new ArrayList<>(filters.size() + 1);
```
Comment (yawkat) [inline http/src/main/java/io/micronaut/http/filter/FilterRunner.java:257]:
```suggestion
        // 1 free spot for the RouteMatchResolverHttpFilter
        List<InternalHttpFilter> filtersToRun = new ArrayList<>(filters.size() + 1);
```

File: http-validation/src/main/java/io/micronaut/validation/routes/FilterVisitor.java [hunk 2]
```diff
@@ -169,6 +170,9 @@ public void visitMethod(MethodElement element, VisitorContext context) {
                     }
                 }
             }
+            if (continuationCreator != null && element.hasStereotype(PreMatching.class)) {
```
Comment (yawkat) [inline http-validation/src/main/java/io/micronaut/validation/routes/FilterVisitor.java]: Maybe also check that this is a ServerFilter

**Review by yawkat — APPROVED** [review]

**Comment by dstepanov** [comment]
@graemerocher The route is resolved after the pre-matching filter so you can simply modify the request

**Comment by graemerocher** [comment]
but the request is not mutable. Does calling `mutate()` work? It would be good to have a test and example of mutating the route

**Comment by dstepanov** [comment]
There are ways to mutate it:
```java
        @Order(1)
        @PreMatching
        @RequestFilter
        MutableHttpRequest<?> preMatchingRequest1(MutableHttpRequest<?> request) {
            return request.uri(new URI(request.getUri().toString().replace("xyz", "HELLO")))
        }

        @Order(2)
        @PreMatching
        @RequestFilter
        void preMatchingRequest2(MutableHttpRequest<?> request) {
            request.uri(new URI(request.getUri().toString().replace("HELLO", "abc")))
        }
  ```

**Comment by dstepanov** [comment]
@yawkat The limitation of using the continuation was in my intial implementation, after the changes it also works in pre-matching

**Review by graemerocher — APPROVED** [review]
Ok, understood. Maybe fix the failing test :)

**Pull request merged**